File: 7.二叉树/11.完全二叉树的节点个数/222.完全二叉树的节点个数.cpp
/*
 * @lc app=leetcode.cn id=222 lang=cpp
 *
 * [222] 完全二叉树的节点个数
 */
#include <bits/stdc++.h>
#include <ranges>
using namespace std;
struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};
// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int countNodes(TreeNode* root) {
        if (! root) return 0;
        TreeNode* tl = root->left;
        TreeNode* tr = root->right;
        int lh = 0, rh = 0;
        while (tl) tl = tl->left, lh ++;
        while (tr) tr = tr->right, rh ++;
        if (lh == rh) return (2 << lh) - 1;
        return countNodes(root->left) + countNodes(root->right) + 1;
    }
};
// @lc code=end

class Solution {
    public:
        int countNodes(TreeNode* root) {
            queue<TreeNode*> q;
            int cnt = 0;
            if (root) q.push(root);
            while (q.size())
            {
                int n = q.size();
                cnt += n;
                while (n --)
                {
                    TreeNode* tmp = q.front();
                    q.pop();
                    if (tmp->left) q.push(tmp->left);
                    if (tmp->right) q.push(tmp->right);
                }
            }
            return cnt;
        }
    };

File: 面经150/面经150p1/9.二叉树/222.完全二叉树的节点个数.py
#
# @lc app=leetcode.cn id=222 lang=python3
#
# [222] 完全二叉树的节点个数
#
from mytools import *
# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def countNodes(self, root: Optional[TreeNode]) -> int:
        def get_h(n):
            h = 0
            while n: h += 1; n = n.left
            return h
        if not root: return 0
        l = get_h(root.left); r = get_h(root.right)
        if l == r: return (1 << l) + self.countNodes(root.right)
        else: return (1 << r) + self.countNodes(root.left)
# @lc code=end

